File: IPWorks SMIME Samples/SMIME/python/smime.py
# 
# IPWorks S/MIME 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks S/MIME in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworkssmime
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworkssmime import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)


smime = SMIME()
certmgr = CertMgr()
certSubjects = []
certCount = -1
inSrc = ""
outSrc = ""

def promptWithDefault(prompt, val):
  temp = input(prompt)
  return val if(temp == "") else temp

def fireCertList(e):
  global certCount, certSubjects
  certCount += 1
  certSubjects = certSubjects[:] + [e.cert_subject]
  print(str(certCount) + ". " + e.cert_subject)

def selectCertificate():
  global certCount
  certCount = -1
  prompt = "Select Certificate Store Type:"
  prompt += "\n  0) PFX File"
  prompt += "\n  1) PEMKey File"
  prompt += "\n  2) Java Key Store (JKS) File"
  prompt += "\nChoice: [0] "
  certTypeChoice = promptWithDefault(prompt, "0")
  if certTypeChoice == "0":
    certmgr.set_cert_store_type(2)
    certStorePath = promptWithDefault("Enter PFXFile Path: [./testcert.pfx] ", "./testcert.pfx")
  elif certTypeChoice == "1":
    certStorePath = promptWithDefault("Enter PEMKeyFile Path: [./testcert.pem] ", "./testcert.pem")
    certmgr.set_cert_store_type(6)
  elif certTypeChoice == "2":
    certStorePath = promptWithDefault("Enter JKS Path: [./testcert.jks] ", "./testcert.jks")
    certmgr.set_cert_store_type(4)
  else:
    print("Not a valid choice. Please try again.")
    selectCertificate()
  
  certmgr.set_cert_store_password(promptWithDefault("Enter store password: [password] ", "password"))
  certmgr.set_cert_store(certStorePath)
  
  print("Please select a certificate:")
  certmgr.list_store_certificates()
  certmgr.set_cert_subject(certSubjects[int(promptWithDefault("\nChoice: [0] ", "0"))])

def doEncrypt():
  getIOSources()
  print("Please select a certificate for encryption.")
  selectCertificate()
  smime.add_recipient_cert(certmgr.get_cert_encoded())
  smime.encrypt()
  
def doDecrypt():
  getIOSources()
  print("Please select a certificate for decryption.")
  selectCertificate()
  
  smime.set_cert_store_type(certmgr.get_cert_store_type())
  smime.set_cert_store(certmgr.get_cert_store())
  smime.set_cert_store_password(certmgr.get_cert_store_password())
  smime.set_cert_subject(certmgr.get_cert_subject())
  smime.decrypt()
  
def doSign():
  getIOSources()
  print("Please select a certificate for signing.")
  selectCertificate()
  smime.set_cert_store_type(certmgr.get_cert_store_type())
  smime.set_cert_store(certmgr.get_cert_store())
  smime.set_cert_store_password(certmgr.get_cert_store_password())
  smime.set_cert_subject(certmgr.get_cert_subject())
  smime.sign()
  
def doVerify():
  getIOSources()
  print("Please select a certificate for verification.")
  selectCertificate()
  smime.set_signer_cert_encoded(certmgr.get_cert_encoded())
  smime.verify_signature()
  
def doSignAndEncrypt():
  getIOSources()
  print("Please select a certificate for signing.")
  selectCertificate()
  smime.set_cert_store_type(certmgr.get_cert_store_type())
  smime.set_cert_store(certmgr.get_cert_store())
  smime.set_cert_store_password(certmgr.get_cert_store_password())
  smime.set_cert_subject(certmgr.get_cert_subject())
  print("Please select a certificate for encryption.")
  selectCertificate()
  smime.add_recipient_cert(certmgr.get_cert_encoded())
  smime.sign_and_encrypt()
  
def doDecryptAndVerify():
  getIOSources()
  print("Please select a certificate for verification.")
  selectCertificate()
  smime.set_signer_cert_encoded(certmgr.get_cert_encoded())
  print("Please select a certificate for decryption.")
  selectCertificate()
  smime.set_cert_store_type(certmgr.get_cert_store_type())
  smime.set_cert_store(certmgr.get_cert_store())
  smime.set_cert_store_password(certmgr.get_cert_store_password())
  smime.set_cert_subject(certmgr.get_cert_subject())
  smime.decrypt_and_verify_signature()

def getIOSources():
  global inSrc, outSrc
  inSrc = promptWithDefault("Read from a file or string? (f/s): ", "f")
  if(inSrc == "s"):
    smime.set_input_message(input("Please enter the input string: "))
  else:
    smime.set_input_file(input("Please enter the input file path: "))
  outSrc = promptWithDefault("Write to a file or string? (f/s): ", "f")
  if(outSrc == "f"):
    smime.set_output_file(input("Please enter the output file path: "))

def main():
  print("\nWelcome to the /n software S/MIME demo.\n")
  prompt = "What would you like to do?"
  prompt += "\n  0) Quit"
  prompt += "\n  1) Encrypt"
  prompt += "\n  2) Decrypt"
  prompt += "\n  3) Sign"
  prompt += "\n  4) Verify"
  prompt += "\n  5) Sign & Encrypt"
  prompt += "\n  6) Decrypt & Verify"
  prompt += "\nChoice: "
  op = input(prompt)
  
  if(op == "0"):
    sys.exit(0)
  elif(op == "1"):
    doEncrypt()
  elif(op == "2"):
    doDecrypt()
  elif(op == "3"):
    doSign()
  elif(op == "4"):
    doVerify()
  elif(op == "5"):
    doSignAndEncrypt()
  elif(op == "6"):
    doDecryptAndVerify()
  else:
    print("Not a valid choice.")
    main()

# By including headers with the output message, 
# it can be passed directly back into this demo.
smime.config("IncludeHeaders=true")
certmgr.on_cert_list = fireCertList
try:
  main()
except IPWorksSMIMEError as e:
  print("An exception occured:\n\n" + e)

if(outSrc == "s"):
  print("OutputMessage: %s\n" %(smime.get_output_message()))
print("\nThe demo has finished running.")

File: IPWorks SMIME Samples/SMIME/php/smime.php
<?php
/*
 * IPWorks S/MIME 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks S/MIME in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworkssmime
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/ipworkssmime_smime.php');
require_once('../include/ipworkssmime_certmgr.php');
require_once('../include/ipworkssmime_const.php');
?>
<?php

function outputUsage() {
  echo "Usage: php smime.php op [options]\n\n";
  echo "Options:\n";
  echo "  -f      the input file\n";
  echo "  -o      the output file (console output is the default)\n";
  echo "  -s      the input message as a string\n";
  echo "  -r      the recipient certificate file for encryption and signature verification (testcert.cer)\n";
  echo "  -c      the private key certificate used for signing and decryption (testcert.pfx)\n";
  echo "  -p      the password for the private key certificate\n";
  echo "  op:     the operation (Encrypt, Decrypt, Sign, Verify, SignAndEncrypt, DecryptAndVerify)\n\n";
  echo "Example: php smime.php encrypt -f test.txt -o test.enc -r testcert.cer\n";
  echo "Example: php smime.php decrypt -f test.enc -o test.txt -c testcert.pfx -p test\n";
}

function input($prompt) {
  echo $prompt;
  $handle = fopen("php://stdin", "r");
  $data = trim(fgets($handle));
  fclose($handle);
  return $data;
}

$smime = new IPWorksSMIME_SMIME();
$smime->setIncludeHeaders(true);

$operation = "";
$recipientCertFile = "";
$certFile = "";
$certPassword = "";
$outputFile = "";
$inputSpecified = false;

if ($argc < 2) {
  outputUsage();
  return;
} else {
  try {
    $operation = $argv[1];
    for ($i = 0; $i < $argc; $i++){
      if (str_starts_with($argv[$i],"-")) {
        if ($argv[$i] == "-f") {
          $smime->setInputFile($argv[$i + 1]);
          $inputSpecified = true;
        } elseif ($argv[$i] == "-o") {
          $outputFile = $argv[$i + 1];
          $smime->setOutputFile($outputFile);
        } elseif ($argv[$i] == "-s") {
          $smime->setInputMessage($argv[$i + 1]);
          $inputSpecified = true;
        } elseif ($argv[$i] == "-r") {
          $recipientCertFile = $argv[$i + 1];
        } elseif ($argv[$i] == "-c") {
          $certFile = $argv[$i + 1];
        } elseif ($argv[$i] == "-p") {
          $certPassword = $argv[$i + 1];
        } else {
          // ignore
        }
      }
    }

    if ($operation == "help" || $operation == "?") {
      outputUsage();
      return;
    }

    if (!$inputSpecified) {
      echo "No input specified. Please set either an input string (-s) or file (-f).";
      return;
    }

    switch (strtolower($operation)) {
      case "encrypt":
        try {
          $smime->setRecipientCertCount(1);
          $smime->setRecipientCertStoreType(0, 99); //auto
          $smime->setRecipientCertStore(0, $recipientCertFile);
          $smime->setRecipientCertSubject(0 , "*");
        } catch (Exception $ex) {
          echo "Error loading certificate: " . $ex->getMessage() . "\n";
          return;
        }
        $smime->doEncrypt();
        echo "Encryption complete.\n";
        break;
      case "decrypt":
        try {
          $smime->setCertStoreType(99); //auto
          $smime->setCertStore($certFile);
          $smime->setCertStorePassword($certPassword);
          $smime->setCertSubject("*");
        } catch (Exception $ex) {
          echo "Error loading certificate: " . $ex->getMessage() . "\n";
          return;
        }
        $smime->doDecrypt();
        echo "Decryption complete.\n";
        break;
      case "sign":
        try {
          $smime->setCertStoreType(99); //auto
          $smime->setCertStore($certFile);
          $smime->setCertStorePassword($certPassword);
          $smime->setCertSubject("*");
        } catch (Exception $ex) {
          echo "Error loading certificate: " . $ex->getMessage() . "\n";
          return;
        }
        // $smime->setDetachedSignature(false); // should we use?
        $smime->doSign();
        echo "Signing complete.\n";
        break;
      case "verify":
        try {
          $smime->setSignerCertStoreType(99); //auto
          $smime->setSignerCertStore($recipientCertFile);
          $smime->setSignerCertSubject("*");
        } catch (Exception $ex) {
          echo "Error loading certificate: " . $ex->getMessage() . "\n";
          return;
        }
        $smime->doVerifySignature();
        echo "Signature verification complete.\n";
        break;
      case "signandencrypt":
        try {
          // encryption cert
          $smime->setRecipientCertCount(1);
          $smime->setRecipientCertStoreType(0, 99); //auto
          $smime->setRecipientCertStore(0, $recipientCertFile);
          $smime->setRecipientCertSubject(0 , "*");
          // signing cert
          $smime->setCertStoreType(99); //auto
          $smime->setCertStore($certFile);
          $smime->setCertStorePassword($certPassword);
          $smime->setCertSubject("*");
        } catch (Exception $ex) {
          echo "Error loading certificate: " . $ex->getMessage() . "\n";
          return;
        }
        $smime->doSignAndEncrypt();
        echo "Signing and Encryption complete.\n";
        break;
      case "decryptandverify":
        try {
          // decryption cert
          $smime->setCertStoreType(99); //auto
          $smime->setCertStore($certFile);
          $smime->setCertStorePassword($certPassword);
          $smime->setCertSubject("*");
          // signature cert (used for signature verification)
          $smime->setSignerCertStoreType(99); //auto
          $smime->setSignerCertStore($recipientCertFile);
          $smime->setSignerCertSubject("*");
        } catch (Exception $ex) {
          echo "Error loading certificate: " . $ex->getMessage() . "\n";
          return;
        }
        $smime->doDecryptAndVerifySignature();
        echo "Decryption and Signature verification complete.\n";
        break;
      default:
        echo "Invalid operation. Valid values are: Encrypt, Decrypt, Sign, Verify, SignAndEncrypt, and DecryptAndVerify\n";
        return;
    }
  } catch (Exeception $ex) {
    echo "Error: " . $ex->getMessage() . "\n";
  }
  if ($outputFile == "") {
    // output the message to the console
    echo "Output message:\n" . $smime->getOutputMessage() . "\n";
  }
}

?>